src/charm.py

```diff
@@ -4,101 +4,424 @@
+
+        ingress_relation_configuration = {
+            "http": {
+                "routers": {},
+                "services": {},
+            }
+        }
+
+        unit_urls = {}
+        # We should cache this, it is very expensive
+        gateway_address = self._get_gateway_address()
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
+        for unit in relation.units:
+            if unit.app is self.app:
+                logger.debug(f"Skipping unit {unit}")
+                continue
 
-        Learn more about Pebble layers at https://github.com/canonical/pebble
+            unit_id = unit.name[len(other_app_name) + 1:]
+
+            # This does not work in CMRs if the application IS NOT ON THE SAME CLUSTER.
+            # We probably could just look up the IP on `relation_joined` using the Kube
+            # API and getting the pod ip.
+            unit_ingress_address = f"{other_app_name}-{unit_id}.{other_app_name}-endpoints.{namespace}.svc.cluster.local"
+
+            traefik_router_name = f"juju-{namespace}-{other_app_name}-{unit_id}-router"
+            traefik_service_name = f"juju-{namespace}-{other_app_name}-{unit_id}-service"
+
+            route_prefix = f"{namespace}-{other_app_name}-{unit_id}"
+            ingress_relation_configuration["http"]["routers"][traefik_router_name] = {
+                "rule": f"PathPrefix(`/{route_prefix}`)",
+                "service": traefik_service_name,
+                "entryPoints": ["web"],
+            }
+
+            ingress_relation_configuration["http"]["services"][traefik_service_name] = {
+                "loadBalancer": {"servers": [{"url": f"http://{unit_ingress_address}:{port}"}]}
+            }
+
+            unit_urls[unit.name] = f"http://{gateway_address}:{self._port}/{route_prefix}"
+
+        ingress_relation_configuration_path = f"{_TRAEFIK_INGRESS_CONFIGURATIONS_DIRECTORY}/{self._ingress_config_file_name(relation)}"
+        container.push(ingress_relation_configuration_path, yaml.dump(ingress_relation_configuration))
+
+        logger.debug(f"Updated ingress configuration file: {ingress_relation_configuration_path}")
+
+        if self.unit.is_leader():
+            # TODO remove this when the SDI library is updated
+            relation.data[self.app]["data"] = yaml.dump({"url": f"http://{gateway_address}:{self._port}/nope", "unit_urls": unit_urls})
+
+        self.unit.status = ActiveStatus()
+
+    def _handle_ingress_broken(self, event: RelationEvent):
+        if not isinstance(self.unit.status, ActiveStatus):
+            logger.debug("Charm not active yet, deferring event")
+            event.defer()
+            return
+
+        relation = event.relation
+
+        container = self.unit.get_container(_TRAEFIK_CONTAINER_NAME)
```

Comment:
This line features a lot throughout this code, you could consider doing

```
self._container = self.unit.get_container(_TRAEFIK_CONTAINER_NAME)
```

In the `__init__`

Reply:
Done in 3cff0be